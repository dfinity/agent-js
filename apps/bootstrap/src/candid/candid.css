:root #candid_ui {
  font-family: 'Circular Std', 'Roboto', sans-serif;
  font-weight: 300;
  font-size: 10px;
  --font-circular: 'Circular Std', 'Roboto', sans-serif;
  --font-mono: 'Vremena Grotesk', 'Roboto Mono', monospace;
  --header-height: 3.8rem;
  --pad-sm: 0.8rem;
  --pad-md: 1.5rem;
  --pad-lg: 2.4rem;
  --font-sm: 1.2rem;
  --font-md: 1.4rem;
  --font-lg: 2.4rem;
  --light-bg: #f6f6f7;
  --white: #f9f9f9;
  --lightest: #ebf0fa;
  --lighter: #eaeaea;
  --light: #e5e5e5;
  --medium: #c3c3c4;
  --dark: #818284;
  --darker: #545454;
  --darkest: #292a2e;
  --error: #b82121;
}

#candid_ui #app {
  display: flex;
  flex-direction: column;
  width: 100vw;
  height: 100vh;
  font-size: var(--font-md);
}

#candid_ui #container {
  display: flex;
  flex-direction: column;
  height: calc(100vh - var(--header-height));
}

#candid_ui #main-content {
  flex: 1;
  overflow: auto;
}

#candid_ui #header {
  height: var(--header-height);
  width: 100%;
  text-align: center;
  font-size: var(--font-md);
  line-height: var(--header-height);
  color: var(--darkest);
  border-bottom: 1px solid var(--dark);
}

#candid_ui #title-card {
  padding: 3.2rem 1.8rem 0;
  font-size: 1.6rem;
}

#candid_ui #title {
  font-size: 3.4rem;
  letter-spacing: 0.1rem;
}

#candid_ui #methods {
  font-family: 'Vremena Grotesk', 'Roboto Mono', monospace;
  padding: 0;
  margin: 0;
}

#candid_ui #methods li {
  display: flex;
  flex-direction: column;
  list-style: none;
  padding: var(--pad-md);
  margin-top: var(--pad-lg);
  border: 1px solid var(--dark);
}

#candid_ui .signature {
  font-weight: lighter;
  margin-bottom: var(--pad-md);
}
.#candid_ui signature b {
  font-weight: bold;
}

#candid_ui .composite {
  margin-bottom: var(--pad-md);
}

#candid_ui input,
#candid_ui select {
  font-size: var(--font-md);
  font-family: 'Circular Std', 'Roboto', sans-serif;
  margin-bottom: var(--pad-lg);
}
#candid_ui input:not([type='checkbox']),
#candid_ui select {
  width: 100%;
  height: 4.8rem;
  padding: 1.5rem;
  border: 1px solid var(--dark);
}
#candid_ui input[type='number']::-webkit-inner-spin-button {
  width: 3rem;
  height: 3rem;
}
#candid_ui label {
  font-family: 'Circular Std', 'Roboto', sans-serif;
  margin-right: var(--pad-sm);
}
#candid_ui label.small {
  font-size: 1rem;
}
#candid_ui .popup-form {
  padding-left: var(--pad-md);
  width: 100%;
}
#candid_ui input[type='checkbox'] ~ .popup-form {
  padding-left: var(--pad-md);
}
#candid_ui .input-container > span > .popup-form,
#candid_ui .ui-result > span > .popup-form {
  padding-left: 0;
}
#candid_ui .input-container span:not(.status),
#candid_ui .ui-result span:not(.status) {
  position: relative;
  overflow-clip: none;
  overflow-y: visible;
}

#candid_ui .reject {
  border-color: var(--error) !important;
  box-shadow: 0 0 1px var(--error);
}
#candid_ui .result {
  display: none;
  position: relative;
  width: 100%;
  margin-top: var(--pad-lg);
  padding-top: 1.5rem;
  font-size: 1.4rem;
  border-top: 1px solid var(--light);
}
#candid_ui .error {
  color: var(--error);
}
#candid_ui .status {
  display: none;
  position: absolute;
  top: 3.1rem;
  left: 0.2rem;
  font-size: 1.2rem;
  color: var(--error);
  white-space: pre;
}
#candid_ui .left {
  flex: 1;
  overflow-wrap: anywhere;
}
#candid_ui .right {
  color: var(--dark);
  align-self: stretch;
  margin-left: var(--pad-md);
}
#candid_ui .clear {
  clear: both;
}

#candid_ui .button-container {
  display: flex;
  margin-top: 0.8rem;
}

#candid_ui .btn {
  font-family: 'Roboto', sans-serif;
  letter-spacing: 0.05rem;
  background-color: var(--darkest);
  color: var(--lightest);
  font-size: 1.4rem;
  padding: 1.5rem;
  border: none;
  text-transform: uppercase;
  flex: 1;
}
#candid_ui .btn:not(:last-child) {
  margin-right: var(--pad-md);
}
#candid_ui button {
  padding: 0;
  border: none;
  background-color: inherit;
  color: inherit;
}
#candid_ui button:focus,
#candid_ui input:focus,
#candid_ui select:focus,
#candid_ui textarea:focus {
  outline: none;
}
#candid_ui .random {
  background-color: var(--lighter);
  color: var(--darkest);
}

#candid_ui .result-buttons {
  display: none;
  position: absolute;
  top: -0.1rem;
  right: 0;
}
#candid_ui .result:not(.error):hover .result-buttons {
  display: block;
}
#candid_ui .result-buttons .btn {
  color: var(--darkest);
  background-color: var(--white);
  border: 1px solid var(--darkest);
  font-size: 1.1rem;
  padding: 0.2rem 0.8rem;
}
#candid_ui .result-buttons .btn:not(:last-child) {
  margin-right: 0.5rem;
}
#candid_ui .result-buttons .btn:focus,
#candid_ui .result-buttons .btn:hover {
  background-color: var(--light);
}
#candid_ui .result-buttons .btn:active {
  background-color: var(--lightest);
}
#candid_ui .result-buttons .btn.active {
  color: var(--white);
  background-color: var(--darkest);
}

#candid_ui #console {
  overflow: hidden;
  height: var(--header-height);
  background-color: var(--light-bg);
  color: var(--darkest);
  transition: height 0.2s ease-in-out;
}
#candid_ui #console.open {
  height: 40vh;
}

#candid_ui #console-bar {
  display: flex;
  height: var(--header-height);
  background-color: var(--white);
  padding: 0;
  border-bottom: 1px solid var(--light);
}
#candid_ui .console-header {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: var(--header-height);
  background-color: var(--light-bg);
  font-size: 2rem;
  font-family: var(--font-circular);
  text-align: center;
  text-transform: uppercase;
}
#candid_ui #output-button,
#candid_ui #methods-button {
  flex: 1;
  cursor: pointer;
  font-size: var(--font-lg);
  border-bottom: 1px solid var(--medium);
}
#candid_ui #output-button:hover,
#candid_ui #methods-button:hover,
#candid_ui #output-button:focus,
#candid_ui #methods-button:focus {
  background-color: var(--light);
}
#candid_ui .active-tab {
  background-color: var(--lighter);
}

#candid_ui #output-pane,
#candid_ui #methods-pane {
  font-family: var(--font-mono);
  height: 0;
  padding: 0 0 3rem;
  margin-right: calc(-1 * var(--pad-md));
}

#candid_ui #output-list,
#candid_ui #methods-list {
  list-style: none;
  padding: 0 0 var(--pad-lg);
  height: calc(40vh - (var(--header-height) * 2));
  overflow-y: auto;
}

#candid_ui .output-line {
  width: 100%;
  padding: 0.4rem 0.8rem;
  overflow-wrap: break-word;
}
#candid_ui .output-line:nth-child(2n-1) {
  background-color: var(--lighter);
}

#candid_ui #methods-list {
  padding-top: 0.2rem;
}
#candid_ui #methods-list a {
  display: block;
  padding: 0.7rem 1.5rem;
  color: var(--darkest);
}
#candid_ui #methods-list a:focus,
#candid_ui #methods-list a:hover {
  background-color: var(--darker);
  color: var(--white);
  text-decoration: none;
}
#candid_ui #methods-list a:active {
  background-color: var(--darkest);
}

@media only screen and (min-width: 800px) {
  :root #candid_ui {
    --header-height: 5.6rem;
    --console-width: clamp(28rem, 30vw, 40rem);
  }

  #candid_ui #header {
    text-align: left;
    padding-left: 3rem;
  }

  #candid_ui #container {
    flex-direction: row;
    overflow: auto;
  }

  #candid_ui #title-card {
    padding-left: 0;
  }

  #candid_ui #main-content {
    flex: 1;
    padding: 0 3rem 5rem;
  }

  #candid_ui .btn {
    width: auto;
    flex: 0;
    padding: var(--pad-sm) var(--pad-md);
  }

  #candid_ui .input-container span:not(:first-child):not(.status),
  #candid_ui .ui-result span:not(:first-child):not(.status) {
    margin-left: var(--pad-md);
  }
  #candid_ui input:not([type='checkbox']),
  #candid_ui select {
    width: 23.2rem;
    height: 4rem;
    padding: 1rem var(--pad-md);
  }

  #candid_ui #console {
    display: flex;
    flex-direction: row-reverse;
    width: var(--header-height);
    height: calc(100vh - var(--header-height));
    padding: 0;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open {
    width: calc(var(--header-height) + var(--console-width));
    height: calc(100vh - var(--header-height));
  }

  #candid_ui #console-bar {
    display: flex;
    flex-direction: column;
    justify-self: end;
    width: var(--header-height);
    height: 100%;
    padding: 0;
    border-bottom: none;
    border-left: 1px solid var(--light);
  }
  #candid_ui #console-bar button {
    flex: 0;
    font-size: 4rem;
    padding: 0.8rem;
  }

  #candid_ui .console-header {
    display: none;
    width: 0;
    border-bottom: none;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open .console-header {
    display: flex;
    width: 100%;
  }

  #candid_ui #output-pane,
  #candid_ui #methods-pane {
    width: 100%;
    margin: 0;
  }

  #candid_ui #methods-pane,
  #candid_ui #output-pane {
    height: 100% !important;
    width: 0;
    transition: width 0.2s ease-in-out;
  }
  #candid_ui #console.open #methods-pane,
  #candid_ui #console.open #output-pane {
    width: var(--console-width);
  }

  #candid_ui #output-list,
  #candid_ui #methods-list {
    height: calc(100vh - (var(--header-height) * 2));
  }
}
