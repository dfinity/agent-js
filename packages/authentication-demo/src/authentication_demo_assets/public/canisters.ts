// @ts-ignore
import authDemoCanister from "ic:canisters/authentication_demo";
import { Actor } from "@dfinity/agent";

// const authentication_demo = Actor.canisterIdOf(authDemoCanister)
export const authentication_demo: Actor = authDemoCanister;
