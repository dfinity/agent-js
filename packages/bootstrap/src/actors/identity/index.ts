export { default as DocumentIdentities } from './DocumentIdentities';
export { default as IdentityActor } from './IdentityActor';
