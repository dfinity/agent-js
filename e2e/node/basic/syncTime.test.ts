import {
  AnonymousIdentity,
  CallRequest,
  HttpAgent,
  IC_REQUEST_DOMAIN_SEPARATOR,
  makeNonce,
  Nonce,
  ReadStateRequest,
  Signature,
  Signed,
  UnSigned,
} from '@dfinity/icp/agent';
import { Principal } from '@dfinity/icp/principal';
import { IDL } from '@dfinity/icp/candid';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { createActor } from '../canisters/counter';
import {
  MockReplica,
  prepareV2ReadStateTimeResponse,
  prepareV3Response,
} from '../utils/mock-replica';
import { randomIdentity, randomKeyPair } from '../utils/identity';
import { concatBytes } from '@noble/hashes/utils';

const INVALID_EXPIRY_ERROR =
  'Invalid request expiry: Specified ingress_expiry not within expected range: Minimum allowed expiry: 2025-05-01 23:55:18.005285297 UTC, Maximum allowed expiry: 2025-05-02 00:00:48.005285297 UTC, Provided expiry: 2025-05-01 12:38:00 UTC';

describe('syncTime', () => {
  const date = new Date('2025-05-01T12:34:56.789Z');
  const canisterId = Principal.fromText('uxrrr-q7777-77774-qaaaq-cai');
  const nonce = makeNonce();

  const ICP_LEDGER = 'ryjl3-tyaaa-aaaaa-aaaba-cai';

  const greetMethodName = 'greet';
  const greetReq = 'world';
  const greetRes = 'Hello, world!';
  const greetArgs = IDL.encode([IDL.Text], [greetReq]);
  const greetReply = IDL.encode([IDL.Text], [greetRes]);

  const keyPair = randomKeyPair();
  const identity = randomIdentity();
  const anonIdentity = new AnonymousIdentity();

  let mockReplica: MockReplica;

  beforeEach(async () => {
    mockReplica = await MockReplica.create();

    vi.useFakeTimers();
    vi.setSystemTime(date);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe('on error', () => {
    it('should not sync time by default', async () => {
      const agent = await HttpAgent.create({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        identity,
      });
      const actor = await createActor(canisterId, { agent });
      const sender = identity.getPrincipal();

      const { responseBody, requestId } = await prepareV3Response({
        canisterId,
        methodName: greetMethodName,
        arg: greetArgs,
        sender,
        reply: greetReply,
        keyPair,
        date,
        nonce,
      });
      const signature = await identity.sign(concatBytes(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(responseBody);
      });

      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);
      expect(actorResponse).toEqual(greetRes);

      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(1);
      expectV3CallRequest(
        mockReplica.getV3CallReq(canisterId.toString(), 0),
        {
          nonce,
          sender,
          pubKey: identity.getPublicKey().toDer(),
          signature,
        },
        'V3 call body',
      );
    });

    it('should sync time when the local time does not match the subnet time', async () => {
      const agent = await HttpAgent.create({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        identity,
      });
      const actor = await createActor(canisterId, { agent });
      const sender = identity.getPrincipal();

      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(400).send(new TextEncoder().encode(INVALID_EXPIRY_ERROR));
      });

      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      const { responseBody: callResponse, requestId } = await prepareV3Response({
        canisterId,
        methodName: greetMethodName,
        arg: greetArgs,
        sender,
        reply: greetReply,
        keyPair,
        date,
        nonce,
      });
      const signature = await identity.sign(concatBytes(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(callResponse);
      });

      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);

      expect(actorResponse).toEqual(greetRes);
      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(2);

      const req = mockReplica.getV3CallReq(canisterId.toString(), 0);
      expectV3CallRequest(
        req,
        {
          nonce,
          sender,
          pubKey: identity.getPublicKey().toDer(),
          signature,
        },
        'V3 call body',
      );

      const reqTwo = mockReplica.getV3CallReq(canisterId.toString(), 1);
      expect(reqTwo).toEqual(req);

      expect(mockReplica.getV2ReadStateSpy(canisterId.toString())).toHaveBeenCalledTimes(3);
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 0),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V2 read state body one',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 1),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body two',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 2),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body three',
      );
    });
  });

  describe('on async creation', () => {
    it('should sync time on when enabled', async () => {
      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(ICP_LEDGER, (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(ICP_LEDGER, (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(ICP_LEDGER, (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      await HttpAgent.create({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        shouldSyncTime: true,
      });

      expect(mockReplica.getV2ReadStateSpy(ICP_LEDGER)).toHaveBeenCalledTimes(3);
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(ICP_LEDGER, 0),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body one',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(ICP_LEDGER, 1),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body two',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(ICP_LEDGER, 2),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body three',
      );
    });

    it('should not sync time by default', async () => {
      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(ICP_LEDGER, (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      await HttpAgent.create({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        shouldSyncTime: false,
        identity: anonIdentity,
      });

      expect(mockReplica.getV2ReadStateSpy(ICP_LEDGER)).toHaveBeenCalledTimes(0);
    });

    it('should not sync time when explicitly disabled', async () => {
      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(ICP_LEDGER, (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      await HttpAgent.create({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        shouldSyncTime: false,
        identity: anonIdentity,
      });

      expect(mockReplica.getV2ReadStateSpy(ICP_LEDGER)).toHaveBeenCalledTimes(0);
    });
  });

  describe('on first call', () => {
    it('should sync time when enabled', async () => {
      const agent = HttpAgent.createSync({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        identity,
        shouldSyncTime: true,
      });
      const actor = await createActor(canisterId, { agent });

      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      const { responseBody, requestId } = await prepareV3Response({
        canisterId,
        methodName: greetMethodName,
        arg: greetArgs,
        sender: identity.getPrincipal(),
        reply: greetReply,
        keyPair,
        date,
        nonce,
      });
      const signature = await identity.sign(concatBytes(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(responseBody);
      });

      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);
      expect(actorResponse).toEqual(greetRes);

      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(1);
      const req = mockReplica.getV3CallReq(canisterId.toString(), 0);
      expectV3CallRequest(
        req,
        {
          nonce,
          sender: identity.getPrincipal(),
          pubKey: identity.getPublicKey().toDer(),
          signature,
        },
        'V3 call body',
      );

      expect(mockReplica.getV2ReadStateSpy(canisterId.toString())).toHaveBeenCalledTimes(3);
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 0),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body one',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 1),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body two',
      );
      expectV2ReadStateRequest(
        mockReplica.getV2ReadStateReq(canisterId.toString(), 2),
        {
          sender: anonIdentity.getPrincipal(),
        },
        'V3 read state body three',
      );
    });

    it('should not sync time by default', async () => {
      const agent = HttpAgent.createSync({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        identity,
      });
      const actor = await createActor(canisterId, { agent });
      const sender = identity.getPrincipal();

      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      const { responseBody, requestId } = await prepareV3Response({
        canisterId,
        methodName: greetMethodName,
        arg: greetArgs,
        sender,
        reply: greetReply,
        keyPair,
        date,
        nonce,
      });
      const signature = await identity.sign(concatBytes(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(responseBody);
      });

      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);
      expect(actorResponse).toEqual(greetRes);

      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(1);
      expectV3CallRequest(
        mockReplica.getV3CallReq(canisterId.toString(), 0),
        {
          nonce,
          sender,
          pubKey: identity.getPublicKey().toDer(),
          signature,
        },
        'V3 call body',
      );

      expect(mockReplica.getV2ReadStateSpy(canisterId.toString())).toHaveBeenCalledTimes(0);
    });

    it('should not sync time when explicitly disabled', async () => {
      const agent = HttpAgent.createSync({
        host: mockReplica.address,
        rootKey: keyPair.publicKeyDer,
        identity,
        shouldSyncTime: false,
      });
      const actor = await createActor(canisterId, { agent });
      const sender = identity.getPrincipal();

      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
        keyPair,
      });
      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(readStateResponse);
      });

      const { responseBody, requestId } = await prepareV3Response({
        canisterId,
        methodName: greetMethodName,
        arg: greetArgs,
        sender,
        reply: greetReply,
        keyPair,
        date,
        nonce,
      });
      const signature = await identity.sign(concatBytes(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
        res.status(200).send(responseBody);
      });

      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);
      expect(actorResponse).toEqual(greetRes);

      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(1);
      const req = mockReplica.getV3CallReq(canisterId.toString(), 0);
      expectV3CallRequest(
        req,
        {
          nonce,
          sender,
          pubKey: identity.getPublicKey().toDer(),
          signature,
        },
        'V3 call body',
      );

      expect(mockReplica.getV2ReadStateSpy(canisterId.toString())).toHaveBeenCalledTimes(0);
    });
  });
});

interface ExpectedV3CallRequest {
  nonce: Nonce;
  sender: Principal;
  pubKey: Uint8Array;
  signature: Signature;
}

function expectV3CallRequest(
  actual: Signed<CallRequest>,
  expected: ExpectedV3CallRequest,
  snapshotName?: string,
) {
  expect(actual.content.nonce).toEqual(expected.nonce);
  expect(actual.content.sender).toEqual(expected.sender.toUint8Array());
  expect(actual.sender_pubkey).toEqual(expected.pubKey);
  expect(actual.sender_sig).toEqual(expected.signature);

  expect(actual).toMatchSnapshot(
    {
      content: {
        nonce: expect.any(Uint8Array),
        sender: expect.any(Uint8Array),
      },
      sender_pubkey: expect.any(Uint8Array),
      sender_sig: expect.any(Uint8Array),
    },
    snapshotName,
  );
}

interface ExpectedV2ReadStateRequest {
  sender: Principal;
}

function expectV2ReadStateRequest(
  actual: UnSigned<ReadStateRequest>,
  expected: ExpectedV2ReadStateRequest,
  snapshotName?: string,
) {
  expect(actual.content.sender).toEqual(expected.sender.toUint8Array());

  expect(actual).toMatchSnapshot(
    {
      content: {
        sender: expect.any(Uint8Array),
      },
    },
    snapshotName,
  );
}
