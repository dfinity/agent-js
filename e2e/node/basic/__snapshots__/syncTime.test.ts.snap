// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncTime > on async creation > should sync time on when enabled > V3 read state body one 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on async creation > should sync time on when enabled > V3 read state body three 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on async creation > should sync time on when enabled > V3 read state body two 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on error > should not sync time by default > V3 call body 1`] = `
{
  "content": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      5,
      119,
      111,
      114,
      108,
      100,
    ],
    "canister_id": Uint8Array [
      0,
    ],
    "ingress_expiry": "1746103080000000000",
    "method_name": "greet",
    "nonce": Any<Uint8Array>,
    "request_type": "call",
    "sender": Any<Uint8Array>,
  },
  "sender_pubkey": Any<Uint8Array>,
  "sender_sig": Any<Uint8Array>,
}
`;

exports[`syncTime > on error > should sync time when the local time does not match the subnet time > V2 read state body one 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on error > should sync time when the local time does not match the subnet time > V3 call body 1`] = `
{
  "content": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      5,
      119,
      111,
      114,
      108,
      100,
    ],
    "canister_id": Uint8Array [
      0,
    ],
    "ingress_expiry": "1746103080000000000",
    "method_name": "greet",
    "nonce": Any<Uint8Array>,
    "request_type": "call",
    "sender": Any<Uint8Array>,
  },
  "sender_pubkey": Any<Uint8Array>,
  "sender_sig": Any<Uint8Array>,
}
`;

exports[`syncTime > on error > should sync time when the local time does not match the subnet time > V3 read state body one 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on error > should sync time when the local time does not match the subnet time > V3 read state body three 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on error > should sync time when the local time does not match the subnet time > V3 read state body two 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on first call > should not sync time by default > V3 call body 1`] = `
{
  "content": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      5,
      119,
      111,
      114,
      108,
      100,
    ],
    "canister_id": Uint8Array [
      0,
    ],
    "ingress_expiry": "1746103080000000000",
    "method_name": "greet",
    "nonce": Any<Uint8Array>,
    "request_type": "call",
    "sender": Any<Uint8Array>,
  },
  "sender_pubkey": Any<Uint8Array>,
  "sender_sig": Any<Uint8Array>,
}
`;

exports[`syncTime > on first call > should not sync time when explicitly disabled > V3 call body 1`] = `
{
  "content": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      5,
      119,
      111,
      114,
      108,
      100,
    ],
    "canister_id": Uint8Array [
      0,
    ],
    "ingress_expiry": "1746103080000000000",
    "method_name": "greet",
    "nonce": Any<Uint8Array>,
    "request_type": "call",
    "sender": Any<Uint8Array>,
  },
  "sender_pubkey": Any<Uint8Array>,
  "sender_sig": Any<Uint8Array>,
}
`;

exports[`syncTime > on first call > should sync time when enabled > V3 call body 1`] = `
{
  "content": {
    "arg": Uint8Array [
      68,
      73,
      68,
      76,
      0,
      1,
      113,
      5,
      119,
      111,
      114,
      108,
      100,
    ],
    "canister_id": Uint8Array [
      0,
    ],
    "ingress_expiry": "1746103080000000000",
    "method_name": "greet",
    "nonce": Any<Uint8Array>,
    "request_type": "call",
    "sender": Any<Uint8Array>,
  },
  "sender_pubkey": Any<Uint8Array>,
  "sender_sig": Any<Uint8Array>,
}
`;

exports[`syncTime > on first call > should sync time when enabled > V3 read state body one 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on first call > should sync time when enabled > V3 read state body three 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;

exports[`syncTime > on first call > should sync time when enabled > V3 read state body two 1`] = `
{
  "content": {
    "ingress_expiry": "1746103080000000000",
    "paths": [
      [
        Uint8Array [
          116,
          105,
          109,
          101,
        ],
      ],
    ],
    "request_type": "read_state",
    "sender": Any<Uint8Array>,
  },
}
`;
